File: Kotlin/src/main/kotlin/parrot/Parrot.kt
package parrot

import kotlin.math.max
import kotlin.math.min

open class Parrot(
    private val type: ParrotTypeEnum,
    protected val numberOfCoconuts: Int,
    protected val voltage: Double,
    protected val isNailed: Boolean
) {

    open val speed: Double
        get() = when (type) {
            ParrotTypeEnum.EUROPEAN -> baseSpeed
            ParrotTypeEnum.AFRICAN -> max(0.0, baseSpeed - loadFactor * numberOfCoconuts)
            ParrotTypeEnum.NORWEGIAN_BLUE -> if (isNailed) 0.0 else getBaseSpeed(voltage)
        }

    protected val loadFactor: Double
        get() = 9.0

    protected val baseSpeed: Double
        get() = 12.0

    protected fun getBaseSpeed(voltage: Double): Double = min(24.0, voltage * baseSpeed)
}

File: Swift/Parrot/Parrot.swift
import Foundation

class Parrot {
    let numberOfCoconuts: Int
    let baseSpeed: Double = 12.0

    init(numberOfCoconuts: Int) {
        self.numberOfCoconuts = numberOfCoconuts
    }

}

class EuropeanParrot: Parrot {
    var speed: Double {
        return baseSpeed
    }
}

class AfricanParrot: Parrot {

    let numberOfCoconutss: Int

    override init(numberOfCoconuts: Int) {
        self.numberOfCoconutss = numberOfCoconuts
        super.init(numberOfCoconuts: numberOfCoconuts)
    }

    var speed: Double {
        let loadFactor: Double = 9.0
        let speedByFactor = loadFactor * Double(numberOfCoconuts)
        return max(0, baseSpeed - speedByFactor)
    }
}

class NorwegianBlueParrot: Parrot {

    let voltage: Double
    let isNailed: Bool

    init(numberOfCoconuts: Int, voltage: Double, isNailed: Bool) {
        self.voltage = voltage
        self.isNailed = isNailed
        super.init(numberOfCoconuts: numberOfCoconuts)
    }

    var speed: Double {
        return (isNailed) ? 0 : baseSpeed(voltage: voltage)
    }
    
    private func baseSpeed(voltage: Double) -> Double {
        return min(24.0, voltage * baseSpeed)
    }
}
